common/include/pcl/common/impl/transforms.hpp

```diff
@@ -37,6 +37,138 @@
+        * \param[out] tgt output 3D point (pointer to 4 floats), can be the same as input. The fourth element is set to 0. */
+      void so3 (const float* src, float* tgt) const
+      {
+        const Scalar p[3] = { src[0], src[1], src[2] };  // need this when src == tgt
+        tgt[0] = static_cast<float> (tf (0, 0) * p[0] + tf (0, 1) * p[1] + tf (0, 2) * p[2]);
+        tgt[1] = static_cast<float> (tf (1, 0) * p[0] + tf (1, 1) * p[1] + tf (1, 2) * p[2]);
+        tgt[2] = static_cast<float> (tf (2, 0) * p[0] + tf (2, 1) * p[1] + tf (2, 2) * p[2]);
+        tgt[3] = 0;
+      }
+
+      /** Apply SE3 transform.
+        * \param[in] src input 3D point (pointer to 3 floats)
+        * \param[out] tgt output 3D point (pointer to 4 floats), can be the same as input. The fourth element is set to 1. */
+      void se3 (const float* src, float* tgt) const
+      {
+        const Scalar p[3] = { src[0], src[1], src[2] };  // need this when src == tgt
+        tgt[0] = static_cast<float> (tf (0, 0) * p[0] + tf (0, 1) * p[1] + tf (0, 2) * p[2] + tf (0, 3));
+        tgt[1] = static_cast<float> (tf (1, 0) * p[0] + tf (1, 1) * p[1] + tf (1, 2) * p[2] + tf (1, 3));
+        tgt[2] = static_cast<float> (tf (2, 0) * p[0] + tf (2, 1) * p[1] + tf (2, 2) * p[2] + tf (2, 3));
+        tgt[3] = 1;
+      }
+    };
+
+#if defined(__SSE2__)
+    /** Optimized version for single-precision transforms using SSE2 intrinsics. */
+    template<>
+    struct Transformer<float>
+    {
+      /// Columns of the transform matrix stored in XMM registers.
+      __m128 c[4];
+
+      Transformer(const Eigen::Matrix4f& tf)
+      {
+        for (size_t i = 0; i < 4; ++i)
+          c[i] = _mm_load_ps (tf.col (i).data ());
+      }
+
+      void so3 (const float* src, float* tgt) const
+      {
+        __m128 p0 = _mm_mul_ps (_mm_load_ps1 (&src[0]), c[0]);
+        __m128 p1 = _mm_mul_ps (_mm_load_ps1 (&src[1]), c[1]);
+        __m128 p2 = _mm_mul_ps (_mm_load_ps1 (&src[2]), c[2]);
+        _mm_store_ps (tgt, _mm_add_ps(p0, _mm_add_ps(p1, p2)));
+      }
+
+      void se3 (const float* src, float* tgt) const
+      {
+        __m128 p0 = _mm_mul_ps (_mm_load_ps1 (&src[0]), c[0]);
+        __m128 p1 = _mm_mul_ps (_mm_load_ps1 (&src[1]), c[1]);
+        __m128 p2 = _mm_mul_ps (_mm_load_ps1 (&src[2]), c[2]);
+        _mm_store_ps (tgt, _mm_add_ps(p0, _mm_add_ps(p1, _mm_add_ps(p2, c[3]))));
+      }
+    };
+
+    /** Optimized version for double-precision transform using SSE2 intrinsics. */
+    template<>
+    struct Transformer<double>
+    {
+      /// Columns of the transform matrix stored in XMM registers.
+      __m128d c[4][2];
+
+      Transformer(const Eigen::Matrix4d& tf)
+      {
+        for (size_t i = 0; i < 4; ++i)
+        {
+          c[i][0] = _mm_load_pd (tf.col (i).data () + 0);
+          c[i][1] = _mm_load_pd (tf.col (i).data () + 2);
+        }
+      }
+
+      void so3 (const float* src, float* tgt) const
```

Comment:
I feel this is misleading. You're taking single precision float pointers at interface level and then treating them as double in the implementation. Why not simply define them double and make equivalent use of `_mm_load_pd1`? According to [these](http://www.univ-orleans.fr/lifo/Members/Sylvain.Jubertie/doc/SIMD/html/group__loadstoreops.html) guys, they're part of the SSE2 intrinsics.

Reply:
But these are point coordinates, e.g. `data[]` member of `PointXYZ`. They are always single-precision.

Reply:
Makes sense :+1: